pkg/resources/validation/roletemplate/roletemplate_test.go

```diff
@@ -0,0 +1,142 @@
+package roletemplate
+
+import (
+	"strconv"
+	"testing"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	controllerv3 "github.com/rancher/webhook/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/mocks"
+	"github.com/stretchr/testify/assert"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var defaultRules = []rbacv1.PolicyRule{
+	{
+		APIGroups: []string{""},
+		Resources: []string{"pods"},
+		Verbs:     []string{"get", "list", "watch"},
+	},
+}
+
+var circleRoleTemplateName = "circleRef"
+
+func TestCheckCircularRef(t *testing.T) {
+	tests := []struct {
+		name           string
+		depth          int
+		circleDepth    int
+		errorDepth     int
+		hasCircularRef bool
+		errDesired     bool
+	}{
+		{
+			name:           "basic test case - no inheritance, no circular ref or error",
+			depth:          0,
+			circleDepth:    -1,
+			errorDepth:     -1,
+			hasCircularRef: false,
+			errDesired:     false,
+		},
+		{
+			name:           "basic inheritance case - depth 1 of input is circular",
+			depth:          1,
+			circleDepth:    0,
+			errorDepth:     -1,
+			hasCircularRef: true,
+			errDesired:     false,
+		},
+		{
+			name:           "self-reference inheritance case - single role template which inherits itself",
+			depth:          0,
+			circleDepth:    0,
+			errorDepth:     -1,
+			hasCircularRef: true,
+			errDesired:     false,
+		},
+		{
+			name:           "deeply nested inheritance case - role template inherits other templates which eventually becomes circular",
+			depth:          3,
+			circleDepth:    2,
+			errorDepth:     -1,
+			hasCircularRef: true,
+			errDesired:     false,
+		},
+		{
+			name:           "basic error case - role inherits another role which doesn't exist",
+			depth:          1,
+			circleDepth:    -1,
+			errorDepth:     0,
+			hasCircularRef: false,
+			errDesired:     true,
+		},
+	}
+
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
```

Comment:
I believe you'd need to capture the range variable just above the call to `Run`, like `testCase := testCase`.

Reply:
I'm not sure what you mean here. I don't really see a need to capture any additional variables - can you explain exactly what you want to see and why?

Reply:
Done - explained offline.